lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -679,6 +688,220 @@ void ComponentOp::getAsmBlockArgumentNames(
+          .getValue();
+  p << " ";
+  p.printSymbolName(componentName);
+
+  // Print the port definition list for input and output ports.
+  auto printPortDefList = [&](auto ports) {
+    p << "(";
+    llvm::interleaveComma(ports, p, [&](const PortInfo &port) {
+      p << "%" << port.name.getValue() << ": " << port.type;
+      if (!port.attributes.empty()) {
+        p << " ";
+        p.printAttributeWithoutType(port.attributes);
+      }
+    });
+    p << ")";
+  };
+  printPortDefList(getInputPortInfo());
+  p << " -> ";
+  printPortDefList(getOutputPortInfo());
+
+  p << " ";
+  p.printRegion(getBody(), /*printEntryBlockArgs=*/false,
+                /*printBlockTerminators=*/false,
+                /*printEmptyBlock=*/false);
+
+  SmallVector<StringRef> elidedAttrs = {"portAttributes", "portNames",
+                                        "portDirections", "sym_name",
+                                        ComponentOp::getTypeAttrName()};
+  p.printOptionalAttrDict((*this)->getAttrs(), elidedAttrs);
+}
+
+ParseResult CombComponentOp::parse(OpAsmParser &parser,
+                                   OperationState &result) {
+  using namespace mlir::function_interface_impl;
+
+  StringAttr componentName;
+  if (parser.parseSymbolName(componentName,
+                             ::mlir::SymbolTable::getSymbolAttrName(),
+                             result.attributes))
+    return failure();
+
+  SmallVector<mlir::OpAsmParser::Argument> ports;
+
+  SmallVector<Type> portTypes;
+  if (parseComponentSignature(parser, result, ports, portTypes))
+    return failure();
+
+  // Build the component's type for FunctionLike trait. All ports are listed
+  // as arguments so they may be accessed within the component.
+  auto type = parser.getBuilder().getFunctionType(portTypes, /*results=*/{});
+  result.addAttribute(CombComponentOp::getTypeAttrName(), TypeAttr::get(type));
+
+  auto *body = result.addRegion();
+  if (parser.parseRegion(*body, ports))
+    return failure();
+
+  if (body->empty())
+    body->push_back(new Block());
+
+  if (parser.parseOptionalAttrDict(result.attributes))
+    return failure();
+
+  return success();
+}
+
+LogicalResult CombComponentOp::verify() {
+  // Verify there is exactly one of each the wires and control operations.
+  auto wIt = getBodyBlock()->getOps<WiresOp>();
+  if (std::distance(wIt.begin(), wIt.end()) != 1)
+    return emitOpError("requires exactly one 'calyx.wires' op.");
+
+  // Verify the component actually does something: has continuous assignments.
+  bool hasNoAssignments =
+      getWiresOp().getBodyBlock()->getOps<AssignOp>().empty();
+  if (hasNoAssignments)
+    return emitOpError(
+        "The component currently does nothing. It needs to either have "
+        "continuous assignments in the Wires region or control constructs in "
+        "the Control region.");
+
+  // Check that all cells are combinational
+  auto cells = getOps<CellInterface>();
+  for (auto cell : cells) {
```

Comment:
Suggestion:
```
if (!llvm::all_of(cells, [](CellInterface cell) { return cell.isCombinational(); }) 
{ 
  return emitOpError() << "..."; 
}
```

Reply:
Ah I guess the downside to the above is you can't determine which is the offending cell. Another suggestion might be:

```
if (auto it = llvm::find_if(cells, [](CellInterface cell) { return !cell.isCombinational(); }); it != cells.end()) {
   ...
} 
```

Reply:
What is the advantage of this new method? Is the way I am currently doing it not correct?

I'm happy to change but your suggestion seems pretty verbose without an obvious advantage.

Reply:
I tend to prefer the functional choice over for loops. They read better as a sentence, and guarantee no OoB accesses (though this applies more so to non-for range loops).

However, up to you - I only marked this as a suggestion.

Reply:
I agree with the functional approach in general, but I think it gets overly complicated in this case.